### PR TITLE
Proposal: deis ps:restart

### PR DESCRIPTION
**Pull request by rjocoleman**
This is a proposal for a `deis` cli command, `ps:restart` with the alias `restart`

`deis ps:restart` should restart an application.

I'll investigate the possibility of only restarting specific container types like the similar Heroku command:

The use case for this command is restarting an app in times of instability without using the existing workaround of the `ps:scale` command.

```
$ heroku help restart
Alias: restart redirects to ps:restart
Usage: heroku ps:restart [DYNO]

 restart an app dyno

 if DYNO is not specified, restarts all dynos on the app

Examples:

 $ heroku ps:restart web.1
 Restarting web.1 dyno... done

 $ heroku ps:restart web
 Restarting web dyno... done

 $ heroku ps:restart
 Restarting dynos... done
```
I'll convert this to a PR and attach docs once I've looked at the relevant code and have feedback that this is a desired function + any opinions on the interface.

Refs #467, #1929, #2758

### CODE REVIEW

**Comment by rjocoleman** [comment]
@carmstrong I've added the controller API docs, I'll need to do the client docs last as they're automagic

File: docs/reference/api-v1.1.rst [hunk 1]
```diff
@@ -407,6 +407,69 @@ Example Response:
...
+    POST /v1/apps/example-go/restart/ HTTP/1.1
+    Host: deis.example.com
+    Content-Type: application/json
+    Authorization: token abc123
+
+Example Response:
+
+.. code-block:: console
+
+    HTTP/1.1 204 NO CONTENT
```
Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: This should be a 200 OK

File: docs/reference/api-v1.1.rst [hunk 2]
```diff
@@ -407,6 +407,69 @@ Example Response:
...
+    POST /v1/apps/example-go/restart/web HTTP/1.1
+    Host: deis.example.com
+    Content-Type: application/json
+    Authorization: token abc123
+
+Example Response:
+
+.. code-block:: console
+
+    HTTP/1.1 204 NO CONTENT
```
Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: again, 200 OK :)

File: docs/reference/api-v1.1.rst [hunk 3]
```diff
@@ -407,6 +407,69 @@ Example Response:
...
+
+
+Restart Application Containers by Type
+``````````````````````````````````````
+
+Example Request:
+
+.. code-block:: console
+
+    POST /v1/apps/example-go/restart/web HTTP/1.1
```
Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: I think `web` should be a query type, so it'd look like `/v1/apps/example-go/restart?procType=web`

File: docs/reference/api-v1.1.rst [hunk 4]
```diff
@@ -407,6 +407,69 @@ Example Response:
...
+
+
+Restart Specific Container
+``````````````````````````
+
+Example Request:
+
+.. code-block:: console
+
+    POST /v1/apps/example-go/restart/web/1 HTTP/1.1
```
Comment (bacongobbler) [inline docs/reference/api-v1.1.rst]: Similar to the former: `/v1/apps/example-go/restart?procType=web&num=1`

**Comment by bacongobbler** [comment]
Hmm, Heroku seems to have two endpoints:

```
DELETE /apps/{app_id_or_name}/dynos/{dyno_id_or_name}
DELETE /apps/{app_id_or_name}/dynos
```

Not sure why they decided on DELETE, but I like the idea of the endpoint being `POST /v1/apps/{app_name}/containers/restart` with query parameters to allow more control over the restart strategy (all, by process type or some other process).

**Comment by johanneswuerbach** [comment]
Maybe because you delete one / all dynos and heroku auto-scales (creates dynos) to the previously specified number.

**Comment by bacongobbler** [comment]
right, that's my thinking too. Just looks awkward when looking at it from a pure REST standpoint (DELETE should mean "delete this dyno permanently", not "delete this dyno, which will be re-created").

**Comment by paralin** [comment]
+1 I like this command :)

**Comment by bacongobbler** [comment]
@rjocoleman are you able to get to this for v1.5? If not that's okay; I can get started on it :)

**Comment by rjocoleman** [comment]
@bacongobbler I've pushed my work in progress, it's by no means complete. I got stuck in the API urls section and de-railed.

I'm flat out with a big project launching at the start of next month so I don't have the capacity to look at this for ~1 month sorry

**Comment by carmstrong** [comment]
> I'm flat out with a big project launching at the start of next month so I don't have the capacity to look at this for ~1 month sorry

No worries. We're going to close this for now, as there have been a lot of changes and this will have to be rebased. If you have time to get back to this and have any questions about its implementation, please ping us and we'll be happy to reopen the PR.

Thanks!

**Pull request closed without merging**